File: algds/codewars/4kyu/all-balanced-parenthesis/rb/parens_v1.rb
##
# Generates all possible combinations of `n` balanced parenthesis.
#
# @param n {Integer} The number of balanced parenthesis.
# @return {String}
#
def parens(n)
  res = []

  f = lambda do |s, l, r|
    return if (l > r)

    res << s if (l.zero? && r.zero?)

    f.call("#{s}(", l - 1, r) if l > 0

    f.call("#{s})", l, r - 1) if r > 0
  end

  f.call('', n, n)

  res
end

File: codewars/4kyu/all-balanced-parenthesis/js/parens_v1.js
const l = console.log.bind(console);

function zero(v) {
  return v === 0;
}

/**
 * Generates all possible combinations of `n` balanced parenthesis.
 *
 * @param {number} n The number of balanced parenthesis.
 * @return {String}
 */
function parens(n) {
  var res = [];

  (function f(s, l, r) {
    if (l > r) return;

    if (zero(l) && zero(r)) return res.push(s);

    if (l > 0) f(`${s}(`, l - 1, r);

    if (r > 0) f(`${s})`, l, r - 1);
  })('', n, n);

  return res;
}

l(parens(2));

export { parens };
